openshift_metrics/utils.py

```diff
@@ -0,0 +1,144 @@
+#   Licensed under the Apache License, Version 2.0 (the "License"); you may
+#   not use this file except in compliance with the License. You may obtain
+#   a copy of the License at
+#
+#        http://www.apache.org/licenses/LICENSE-2.0
+#
+#   Unless required by applicable law or agreed to in writing, software
+#   distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+#   WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+#   License for the specific language governing permissions and limitations
+#   under the License.
+#
+
+import datetime
+import json
+import requests
+import time
+
+import openshift as oc
+
+
+def query_metric(openshift_url, token, metric, report_date, retry=3):
+    attempt = 0
+    data = None
+    headers = {'Authorization': "Bearer %s" % token}
+    day_url_vars = "start=%sT00:00:00Z&end=%sT23:59:59Z" % (report_date, report_date)
+    print("Retrieving metric: %s" % metric)
+    while attempt < retry:
+        url = "%s/api/v1/query_range?query=%s&%s&step=60s" % (openshift_url, metric, day_url_vars)
+        r = requests.get(url, headers=headers, verify=False)
+        if r.status_code != 200:
+            print("%s Response: %s" % (r.status_code, r.reason))
+        else:
+            data = r.json()['data']['result']
+            if data:
+                break
+            else:
+                print("Empty result set")
+        attempt = attempt + 1
+        time.sleep(3)
+    if not data:
+        raise Exception('Error retrieving metric: %s' % metric)
+    return data
+
+def get_namespace_annotations():
+    namespaces_dict = {}
+    namespaces = oc.selector("namespaces").objects()
+    for namespace in namespaces:
+        namespace_dict = json.loads(namespace.as_json())['metadata']
```

Comment:
You don't need to convert your structured data into JSON and then parse it back into structured data. You can just write:

```
namespace_dict = namespace.as_dict()
```

If you're looking for *specific* annotations, you don't even need to convert it into a `dict` first; you can use the `get_annotation` method.

Reply:
In retrospect what I had there was somewhat bizarre. I'm going to keep the generic annotations dict for now, as I suspect that it's possible that we end up wanting to retrieve more annotations.